### PR TITLE
Add retain, dup, qos to MQTT on-message callback.

### PR DESCRIPTION
**Pull request by graebm**
This is a backwards compatible change.

Mark existing `OnPublishReceivedHandler` deprecated, replace with `OnMessageReceivedHandler`. New handler carries 3 more args: `retain`, `qos`, and `dup`. Mark functions that take old handler as deprecated, and add otherwise identical function that accept new handler.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by sdavtaker — CHANGES_REQUESTED** [review]
Added a few variablel renaming comments for readability.

File: source/mqtt/MqttClient.cpp [hunk 1]
```diff
@@ -427,6 +431,15 @@ namespace Aws
             }
 
             bool MqttConnection::SetOnMessageHandler(OnPublishReceivedHandler &&onPublish) noexcept
+            {
+                return SetOnMessageHandler(
+                    [onPublish](
```
Comment (bretambrose) [inline source/mqtt/MqttClient.cpp:436]: Is the capture here correct?  Memory is fuzzy wrt capture movable type.

File: source/mqtt/MqttClient.cpp [hunk 2]
```diff
@@ -66,7 +66,7 @@ namespace Aws
                 PubCallbackData() : connection(nullptr), allocator(nullptr) {}
 
                 MqttConnection *connection;
-                OnPublishReceivedHandler onPublishReceived;
+                OnMessageReceivedHandler onPublishReceived;
```
[the remark this answers is not in the document]
Comment (graebm) [reply source/mqtt/MqttClient.cpp:69]: Good call. Didn't do it before just to keep the diff as simple as possible.
But yeah, makes the code more understandable going forward from here

File: source/mqtt/MqttClient.cpp (hunk 1 above)
Comment (graebm) [reply source/mqtt/MqttClient.cpp:436]: Yeah, it just makes a copy of onPublish.

If we were using c++14, we could do a std::move when capturing, but alas

**Review by sdavtaker — APPROVED** [review]

**Pull request merged**